stl/inc/condition_variable

```diff
@@ -214,6 +214,7 @@ public:
             _Unlock_guard<_Lock> _Unlock_outer{_Lck};
             const auto _Now = _Clock::now();
             if (_Now >= _Abs_time) {
+                _Guard.unlock();

```

Comment:
It seems scary to me that this could end up calling _Cnd_timedwait without holding _Lck; the proofs of correctness of condition_variable_any are based on that strict ordering; this form may create missed wakes?

Can we `return _Pred()` here rather than `break`?

Reply:
I don't think this change has any effect because _Guard is already destroyed by the `break;`.

Reply:
Wow, I'm putting my foot in my mouth a lot today. I think this is subtle enough to deserve a comment that this is reordering _Unlock_outer and _Guard's destruction.

And I go back to my "I'm not convinced that this actually meets the condition_variable guarantees on the condition_variable_any" because the modification order on the cv_any is provided by that internal mutex, and we're not following that total order anymore.

But I've already put my foot in my mouth several times, so take with mountains of salt. If this is a correct fix it deserves a comment.

Reply:
The fix is that we need to drop the internal mutex (`*_Myptr`) before acquiring the external lock (`_Lck`) which otherwise happens in the wrong order since `_Unlock_outer` and `_Guard` are destroyed in reverse construction order.

Reply:
> The fix is that we need to drop the internal mutex (`*_Myptr`) before acquiring the external lock (`_Lck`) which otherwise happens in the wrong order since `_Unlock_outer` and `_Guard` are destroyed in reverse construction order.

Right, but if you do that, are the total order semantics of condition variables still met? The proof of a total order is trivial before because anything significant happens under a mutex, and mutexes enforce a total order. This is moving work out of that mutex so we need to prove that external callers who call `notify_all` while holding the external mutex still get correct behavior (no missed wakes).

If destruction order does the wrong thing perhaps RAII is not the right tool here; for example another path that could result in the same deadlock if this is true is Clock::now throwing.

Reply:
(To be clear, I'm not saying that it is wrong. Just that the road of CVs is paved with a lot of incorrect behavior :) )

Reply:
I've added a comment, following @AlexGuteniev's PR title.

The approach seems correct to me, especially because the other control flow already does this:

https://github.com/microsoft/STL/blob/dc888f7d9fb7a4db8d3441f9e9bac2e0c6ecc4db/stl/inc/condition_variable#L226-L235

I don't see a better way to get RAII to do this for us, so a comment seems sufficient.

(As an aside, I have no objection to the lack of added tests here.)